bin/service-mango-pnl/src/main.rs

```diff
@@ -0,0 +1,329 @@
+use {
+    log::*,
+    mango_feeds_connector::chain_data::ChainData,
+    serde_derive::{Deserialize, Serialize},
+    solana_sdk::pubkey::Pubkey,
+    std::str::FromStr,
+    std::{
+        fs::File,
+        io::Read,
+        mem::size_of,
+        sync::{atomic::AtomicBool, Arc, RwLock},
+        time::Duration,
+    },
+};
+
+use anchor_client::Cluster;
+use anchor_lang::Discriminator;
+use fixed::types::I80F48;
+use mango_feeds_connector::metrics::*;
+use mango_v4::state::{MangoAccount, MangoAccountValue, PerpMarketIndex};
+use mango_v4_client::{
+    chain_data, health_cache, AccountFetcher, Client, MangoGroupContext, TransactionBuilderConfig,
+};
+use solana_sdk::commitment_config::CommitmentConfig;
+use solana_sdk::{account::ReadableAccount, signature::Keypair};
+#[derive(Clone, Debug, Deserialize)]
+pub struct PnlConfig {
+    pub update_interval_millis: u64,
+    pub mango_program: String,
+    pub mango_group: String,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct JsonRpcConfig {
+    pub bind_address: String,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct Config {
+    pub source: SourceConfig,
+    pub snapshot_source: SnapshotSourceConfig,
+    pub metrics: MetricsConfig,
+    pub pnl: PnlConfig,
+    pub jsonrpc_server: JsonRpcConfig,
+}
+
+type PnlData = Vec<(Pubkey, Vec<(PerpMarketIndex, I80F48)>)>;
+
+async fn compute_pnl(
+    context: Arc<MangoGroupContext>,
+    account_fetcher: Arc<impl AccountFetcher>,
+    account: &MangoAccountValue,
+) -> anyhow::Result<Vec<(PerpMarketIndex, I80F48)>> {
+    let health_cache = health_cache::new(&context, account_fetcher.as_ref(), account).await?;
+
+    let pnls = account
+        .active_perp_positions()
+        .filter_map(|pp| {
+            if pp.base_position_lots() != 0 {
+                return None;
+            }
+            let pnl = pp.quote_position_native();
+            let settle_token_index = context
+                .perp_markets
+                .get(&pp.market_index)
+                .unwrap()
+                .market
+                .settle_token_index;
+            let perp_settle_health = health_cache.perp_max_settle(settle_token_index).unwrap();
```

Comment:
background on `.unwrap`:
* `self.token_info_index(settle_token_index)?` should never fail because the token must be available in map
* spot_amount_taken_for_health_zero never returns Err